File: library/merge_sort.cpp
// 分割統治法の有名問題

#include <iostream>
#include <algorithm>
using namespace std;
#define MAX_N 500001
#define INF 1000000000000

long int num = 0;
long int L[MAX_N];
long int R[MAX_N];

void merge(int N, long int *X, int left, int mid, int right){
    int n1 = mid - left + 1;
    int n2 = right - mid;
    for (int i = 0; i < n1; i++){L[i] = X[left + i];}
    for (int i = 0; i < n2; i++){R[i] = X[mid + 1 + i];}
    L[n1] = INF; R[n2] = INF;
    int l = 0;
    int r = 0;
    for (int i = 0; i < n1 + n2; i++){
        if (L[l] < R[r]){
            X[left + i] = L[l];
            l++;
        }else{
            X[left + i] = R[r];
            r ++;
        }
        num += 1;
    }
}

void merge_sort(int N, long int *X, int left, int right){
    if (left < right){
        int mid = (left + right)/2;
        merge_sort(N, X, left, mid);
        merge_sort(N, X, mid + 1, right);
        merge(N, X, left, mid, right);
    }
}

int main(){
    int N;
    long int X[MAX_N];
    cin >> N;
    for (int i = 0; i < N; i++){
        cin >> X[i];
    }
    merge_sort(N, X, 0, N-1);
    for (int i = 0; i < N; i++){
        cout << X[i];
        if (i != N - 1){
            cout << " ";
        }
    }
    cout << endl;
    cout << num << endl;
    return 0;
}

File: library/merge_sort.py
import numpy as np
import math

INF = 1000000000000
count__ = 0

def merge(n, S, left, mid, right):
    L = [S[i] for i in range(left, mid + 1)] #size = mid - left + 1
    R = [S[i] for i in range(mid + 1, right + 1)] #size = right - mid
    l = 0
    r = 0
    L.append(INF)
    R.append(INF)
    cnt = 0
    while(cnt < right - left + 1):
        if (L[l] < R[r]):
            S[cnt + left] = L[l]
            l += 1
        else:
            S[cnt + left] = R[r]
            r += 1
        cnt += 1
        global count__
        count__ += 1
    return S

def merge_sort(n: int, S: list, left: int, right: int):
    if (left == right):
        pass
    else:
        mid = (left + right) // 2
        merge_sort(n, S, left, mid)
        merge_sort(n, S, mid + 1, right)
        S = merge(n, S, left, mid, right)
    return S

def main():
    n = int(input())
    S = [int(c) for c in input().split()]
    S = merge_sort(n, S, 0, n - 1)
    for i in range(len(S)):
        print(S[i], end="")
        if i != len(S) - 1:
            print(end=" ")
    print()
    print(count__)


if __name__ == '__main__':
    main()